### ISSUE TITLE
False detection?

### ISSUE THREAD

**Issue by sm9cc** [SpiderLabs/owasp-modsecurity-crs#682]
Hello, I use IPB with SEO friendly URLs and the following rule is getting triggered.

`SecRule REQUEST_URI|REQUEST_HEADERS|ARGS|ARGS_NAMES "@validateByteRange 1-255" \
  "phase:request,\
   rev:'2',\
   ver:'OWASP_CRS/3.0.0',\
   maturity:'9',\
   accuracy:'9',\
   block,\
   msg:'Invalid character in request (null character)',\
   id:920270,\
   severity:'CRITICAL',\
   t:none,t:urlDecodeUni,\
   tag:'application-multi',\
   tag:'language-multi',\
   tag:'platform-multi',\
   tag:'attack-protocol',\
   tag:'OWASP_CRS/PROTOCOL_VIOLATION/EVASION',\
   setvar:'tx.msg=%{rule.msg}',\
   setvar:tx.anomaly_score=+%{tx.error_anomaly_score},\
   setvar:tx.%{rule.id}-OWASP_CRS/PROTOCOL_VIOLATION/EVASION-%{matched_var_name}=%{matched_var}"`

`{
    "transaction": {
        "client_ip": "",
        "time_stamp": "Wed Feb  1 15:33:38 2017",
        "server_id": "0c6d24e2850275f97a9fe8959eabfec453959494",
        "client_port": 0,
        "host_ip": "",
        "host_port": 443,
        "id": "148595961899.173235",
        "request": {
            "protocol": "GET",
            "http_version": 1.1,
            "uri": "/topic/2216-pixel%E2%84%A2-roadtomge-1-week-rofls-punishment-appeal/?do=getNewComment",
            "headers": {
                "Host": "www.fragdeluxe.com",
                "Connection": "Keep-Alive",
                "Accept-Encoding": "gzip",
                "CF-IPCountry": "GB",
                "X-Forwarded-For": "",
                "CF-RAY": "32a61ea3192d3542-LHR",
                "X-Forwarded-Proto": "https",
                "CF-Visitor": "{\"scheme\":\"https\"}",
                "cache-control": "max-age=0",
                "upgrade-insecure-requests": "1",
                "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.87 Safari/537.36",
                "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
                "accept-language": "en-US,en;q=0.8",
                "cookie": "__cfduid=db354ab4c9ebd4ac0568faafeb05758061485905541; ips4_hasJS=true; ips4_IPSSessionFront=tq1n704k6j0bud12choupdkda7; __utmt=1; _gat=1; ips4_member_id=1; ips4_pass_hash=046e5d744a5763d01066a0a913038479; ips4_updateBannerDismiss=true; ips4_licenseDismiss=true; ips4_ipsTimezone=Europe/London; _ga=GA1.2.1353499801.1485905609; __utma=116704593.1353499801.1485905609.1485905609.1485959093.2; __utmb=116704593.24.10.1485959093; __utmc=116704593; __utmz=116704593.1485905609.1.1.utmcsr=(direct)|utmccn=(direct)|utmcmd=(none)",
                "CF-Pseudo-IPv4": "",
                "CF-Connecting-IPv6": "",
                "CF-Connecting-IP": ""
            }
        },
        "response": {
            "http_code": 403,
            "headers": {
                "Content-Encoding": "gzip",
                "Connection": "keep-alive",
                "Server": "Powered By FragDeluxe",
                "Server": "Powered By FragDeluxe",
                "Date": "Wed, 01 Feb 2017 14:33:38 GMT",
                "Content-Type": "text/html",
                "Vary": "Accept-Encoding",
                "Strict-Transport-Security": "max-age=31536000; includeSubdomains",
                "X-Powered-By": "FragDeluxe",
                "X-Frame-Options": "SAMEORIGIN",
                "X-Content-Type-Options": "nosniff",
                "X-Xss-Protection": "1; mode=block"
            }
        },
        "producer": {
            "modsecurity": "ModSecurity v3.0.0-alpha (Linux)",
            "connector": "ModSecurity-nginx v0.1.1-beta",
            "secrules_engine": "Enabled",
            "components": [
                "OWASP_CRS/3.0.0\""
            ]
        },
        "messages": [
            {
                "message": "Invalid character in request (null character)",
                "details": {
                    "match": "Matched \"Operator `@validatebyterange' with parameter `1-255' against variable `REQUEST_URI' (Value: `/topic/2216-pixel\\xffffffe2\\xffffff84\\xffffffa2-roadtomge-1-week-rofls-punishment-appeal/?do=getNewC (6 characters omitted)' ) \" at REQUEST_URI",
                    "ruleId": "920270",
                    "file": "/etc/nginx/rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf",
                    "lineNumber": "490",
                    "data": "",
                    "severity": "2",
                    "ver": "OWASP_CRS/3.0.0",
                    "rev": "2",
                    "tags": [
                        "application-multi",
                        "language-multi",
                        "platform-multi",
                        "attack-protocol",
                        "OWASP_CRS/PROTOCOL_VIOLATION/EVASION"
                    ],
                    "maturity": "9",
                    "accuracy": "9"
                }
            }
        ]
    }
}
`

You can see here: https://puu.sh/tJFQd/8f836c678f.png

I am not sure if its a false positive.

Thanks.

**Comment by dune73** [comment]
@SM91337, thank you for reporting. I confirm, this is a false positive due to the innovative use of hex-encoded chars in the URI.

Are you able to handle the FP yourself, or do you need support?

Also not sure, this is a general FP, we should fight, or just a rare edge case. More input welcome.

**Comment by sm9cc** [comment]
Hi, I have unfortunately disabled mod security for the time being, I felt it was too much to manually fix the false positives, I will be looking to try it again in the future,

Thanks for your confirmation though.

**Comment by dune73** [comment]
What Paranoia Level are you working on? On the default PL1 there should not be too many. If there are, then we would really like to know. The one above is a bit of an edge case.

In case you are not familiar with my False Positive tutorial: https://www.netnea.com/cms/apache-tutorial-8_handling-false-positives-modsecurity-core-rule-set/
People report it really simplifies the process. There is even a script that helps you generate false positives on the fly based on the error log.

**Comment by dune73** [comment]
This is very exotic and we won't fix for the 3.0 release line. Maybe 3.1. Tagging it that way.

**Comment by csanders-git** [comment]
I'm not able to replicate this issue. If the issue does exist it is probably due to a transformation, but all i'm seeing is unidecode. The only way i'm aware of that a server will process a null char in a URL is due to the use of %00. The Hex encoding doesn't seem to be decoded by our transformation and there is no indication of a null byte. 
The code for urldecodeuni is available here (https://github.com/SpiderLabs/ModSecurity/blob/923c3c67938da4de4f7f147816b6d2d6ffff5e6f/apache2/msc_util.c#L1583) it is looking just for a percent sign. 

tl;dr: can't replicate this issue, don't think the hex encoding is causing the problem because it is just treated as ASCII by the rule.

**Comment by fgsch** [comment]
This issue has timed out as it has not received any update in over 2 years.
If this is still a problem please open a new issue.
